src/processors/universal.ts

```diff
@@ -370,6 +360,108 @@ class UniversalProcessor extends BaseOrderProcessor {
+    if (this.isUnlockBatchLocked) {
+      setTimeout(() => {
+        this.unlockOrder(orderId, order, context);
+      }, 0);
+      return;
+    }
+    const logger = context.logger;
+    this.isUnlockBatchLocked = true;
+    try {
+      logger.info(`Unlocking is started`);
+      const key = `${order.give.chainId}_${order.take.chainId}`;
+      let orderIds = this.unlockBatchesOrderIdMap.get(key);
+      if (!orderIds) {
+        orderIds = new Set();
+      }
+      orderIds.add(orderId);
+      this.unlockBatchesOrderIdMap.set(key, orderIds);
+
+      if (orderIds.size !== this.UNLOCK_BATCH_SIZE) {
+        logger.debug(`Batch for unlock is not full`);
+        return;
+      }
+
+      const fees = await this.getFee(order, context);
+      const executionFeeAmount = await context.config.client.getAmountToSend(
+        order.take.chainId,
+        order.give.chainId,
+        fees.executionFees.total,
+        this.context.takeChain.fulfullProvider.connection as Web3
+      );
+      logger.debug(`executionFeeAmount = ${executionFeeAmount.toString()}`);
+      const beneficiary = context.giveChain.beneficiary;
+      if (
+        order.give.chainId === ChainId.Solana ||
+        order.take.chainId === ChainId.Solana
+      ) {
+        for (const orderId of orderIds) {
+          try {
+            const unlockTx = await this.createOrderUnlockTx(
+              orderId,
+              order,
+              beneficiary,
+              executionFeeAmount,
+              fees,
+              context,
+              logger
+            );
+
+            const txUnlock =
+              await this.context.takeChain.unlockProvider.sendTransaction(
+                unlockTx,
+                {
+                  logger,
+                }
+              );
+            logger.info(`unlock transaction ${txUnlock} is completed`);
+          } catch (e) {
+            logger.error("Error in unlocking");
+          }
+        }
+      } else {
+        const batchUnlockTx = await context.config.client.sendBatchUnlock(
+          Array.from(orderIds),
+          order.give.chainId,
+          order.take.chainId,
+          beneficiary,
+          executionFeeAmount,
+          {
+            web3: (this.context.takeChain.unlockProvider as EvmProviderAdapter)
+              .connection,
+            loggerInstance: createClientLogger(logger),
+            reward1: 0,
+            reward2: 0,
+          }
+        );
+
+        const txUnlock =
+          await this.context.takeChain.unlockProvider.sendTransaction(
+            batchUnlockTx,
+            {
+              logger,
+            }
+          );
+
+        logger.info(
+          `unlock for ${JSON.stringify(
+            Array.from(orderIds)
+          )} orders ${txUnlock} is completed`
+        );
+      }
+      this.unlockBatchesOrderIdMap.set(key, new Set());
+    } catch (e) {
+      logger.error(`Error in execution batch unlock ${e}`);
+    }
+    this.isUnlockBatchLocked = false;
```

Comment:
After batch-unlock if finished, the code must check all other queues: probably, it must start unlocking again to this or other chains